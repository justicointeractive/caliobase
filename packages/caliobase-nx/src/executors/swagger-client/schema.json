{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "SwaggerClient executor",
  "description": "",
  "type": "object",
  "properties": {
    "input": {
      "type": "string",
      "description": "path to an openapi/swagger spec file"
    },
    "output": {
      "type": "string",
      "description": "path to the directory where the generated client will be saved"
    },
    "generateSpecTarget": {
      "type": "string",
      "description": "a target to be executed that should generate a swagger file"
    },
    "dotenv": {
      "type": "string",
      "description": "path to a dotenv file to load vars from"
    }
  },
  "required": []
}
