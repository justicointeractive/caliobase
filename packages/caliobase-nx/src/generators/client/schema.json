{
  "$schema": "http://json-schema.org/schema",
  "$id": "Client",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "apiProjectName": {
      "type": "string",
      "description": "",
      "x-prompt": "Which project provides the backend api?",
      "x-dropdown": "projects"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed",
      "alias": "d"
    }
  },
  "required": [
    "name",
    "apiProjectName"
  ]
}
