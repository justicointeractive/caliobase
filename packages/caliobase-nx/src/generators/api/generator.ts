import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  Tree,
} from '@nrwl/devkit';
import { applicationGenerator } from '@nrwl/nest/src/generators/application/application';
import * as path from 'path';
import { ApiGeneratorSchema } from './schema';

interface NormalizedSchema extends ApiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ApiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

export default async function (tree: Tree, options: ApiGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, { name: options.name });
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    normalizedOptions.projectRoot,
    {
      cmsProjectName: null,
      ...options,
      ...normalizedOptions,
      tmpl: '',
    }
  );
  await formatFiles(tree);
}
