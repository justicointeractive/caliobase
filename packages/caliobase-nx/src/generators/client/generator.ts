import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nx/devkit';
import { libraryGenerator } from '@nx/js/src/generators/library/library';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { ClientGeneratorSchema } from './schema';

interface NormalizedSchema extends ClientGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ClientGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).libsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    tmpl: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: ClientGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await libraryGenerator(tree, {
    name: options.name,
    directory: options.directory,
    skipFormat: true,
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    config.targets = {
      ...config.targets,
      'generate-sources': {
        executor: '@caliobase/caliobase-nx:swagger-client',
        outputs: [`${normalizedOptions.projectRoot}/src`],
        options: {
          input: `./${normalizedOptions.projectRoot}/assets/openapi.json`,
          output: `./${normalizedOptions.projectRoot}/src/lib`,
          name: `${normalizedOptions.projectName}.ts`,
        },
        dependsOn: [{ projects: 'dependencies', target: 'swagger' }],
      },
    };
    config.implicitDependencies = [
      ...(config.implicitDependencies ?? []),
      options.apiProjectName,
    ];

    return config;
  });
  modifyProjectConfiguration(
    tree,
    options.apiProjectName,
    (apiProjectConfig) => {
      apiProjectConfig.targets = {
        ...apiProjectConfig.targets,
        swagger: {
          executor: 'nx:run-commands',
          outputs: [`${normalizedOptions.projectRoot}/assets`],
          dependsOn: [
            {
              target: 'build',
              projects: 'self',
            },
          ],
          options: {
            command: `node ./dist/${apiProjectConfig.root}/main.js --write-swagger-and-exit=./${normalizedOptions.projectRoot}/assets/openapi.json`,
          },
        },
      };

      return apiProjectConfig;
    }
  );

  // TODO: update nx.json's targetDefaults build dependsOn "^generate-sources", "generate-sources"

  await formatFiles(tree);
}
