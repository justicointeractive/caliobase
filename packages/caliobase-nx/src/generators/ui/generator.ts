import {
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { applicationGenerator } from '@nrwl/react';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { addDependencyVersionsToPackageJson } from '../../lib/versions';
import { UiGeneratorSchema } from './schema';
import assert = require('assert');

interface NormalizedSchema extends UiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: UiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: UiGeneratorSchema) {
  const tasks: GeneratorCallback[] = [];
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, {
    ...options,
    e2eTestRunner: 'cypress',
    linter: Linter.EsLint,
    unitTestRunner: 'jest',
    skipFormat: true,
    style: 'css',
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    assert(config.targets);
    config.targets['serve'].options[
      'proxyConfig'
    ] = `${normalizedOptions.projectRoot}/proxy.conf.json`;
    return config;
  });
  await formatFiles(tree);

  tasks.push(
    addDependencyVersionsToPackageJson(tree, ['@caliobase/caliobase-ui'])
  );

  return runTasksInSerial(...tasks);
}
