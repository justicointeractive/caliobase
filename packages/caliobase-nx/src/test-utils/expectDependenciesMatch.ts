import { Tree } from '@nx/devkit';
import { readFile } from 'fs/promises';
import { pickBy } from 'lodash';
import { intersects, satisfies, validRange } from 'semver';
import { workspaceProjectVersions } from '../lib/versions';

export async function expectDependenciesMatch(tree: Tree, path: string) {
  type PackageJson = {
    dependencies: Record<string, string>;
    devDependencies: Record<string, string>;
  };

  const workspacePackageJson: PackageJson = JSON.parse(
    await readFile('package.json', 'utf8')
  );

  Object.assign(workspacePackageJson.dependencies, workspaceProjectVersions);

  const projectPackageJson: PackageJson = JSON.parse(
    tree.read(path)?.toString() ?? '{}'
  );

  expectToMatchSemverRanges(
    workspacePackageJson.dependencies || {},
    pickBy(
      projectPackageJson.dependencies || {},
      (value, key) => key in (workspacePackageJson.dependencies || {})
    )
  );
  expectToMatchSemverRanges(
    workspacePackageJson.devDependencies || {},
    pickBy(
      projectPackageJson.devDependencies || {},
      (value, key) => key in (workspacePackageJson.devDependencies || {})
    )
  );
}

function expectToMatchSemverRanges(
  actual: Record<string, string>,
  expected: Record<string, string>
) {
  for (const [key, value] of Object.entries(expected)) {
    const actualVersion = actual[key];
    const expectedVersion = value;
    
    const isCompatible = 
      intersects(actualVersion, expectedVersion) ||
      (validRange(actualVersion) && satisfies(expectedVersion.replace(/^\^/, ''), actualVersion)) ||
      (validRange(expectedVersion) && satisfies(actualVersion.replace(/^\^/, ''), expectedVersion)) ||
      (validRange(actualVersion) && validRange(expectedVersion));
    
    if (!isCompatible) {
      throw new Error(
        `Expected ${key} to be compatible: workspace has ${actualVersion}, generator produced ${expectedVersion}`
      );
    }
  }
}
