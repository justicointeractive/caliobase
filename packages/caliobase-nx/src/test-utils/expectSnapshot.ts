import { Tree } from '@nrwl/devkit';

export function expectSnapshot(tree: Tree, path: string) {
  expect(tree.read(path)?.toString()).toMatchSnapshot(path);
}
