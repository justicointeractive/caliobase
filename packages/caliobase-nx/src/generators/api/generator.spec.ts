import { readProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { expectDependenciesMatch } from '../../test-utils/expectDependenciesMatch';
import { expectSnapshot } from '../../test-utils/expectSnapshot';

import generator from './generator';
import { ApiGeneratorSchema } from './schema';

describe('api generator', () => {
  let appTree: Tree;
  const options: ApiGeneratorSchema = { name: 'api' };

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await generator(appTree, options);
    const config = readProjectConfiguration(appTree, 'api');
    expect(config).toBeDefined();

    expectSnapshot(appTree, 'package.json', [
      [/"@caliobase\/caliobase": "[0-9.]+"/g, '"@caliobase/caliobase": "*"'],
    ]);
    await expectDependenciesMatch(appTree, 'package.json');
    expectSnapshot(appTree, 'apps/api/project.json');
    expectSnapshot(appTree, 'apps/api/src/main.ts');
    expectSnapshot(appTree, 'apps/api/src/app/app.module.ts');
    expectSnapshot(
      appTree,
      'apps/api/src/app/entities/organization-profile.entity.ts'
    );
    expectSnapshot(appTree, 'apps/api/src/app/entities/user-profile.entity.ts');
  });
});
