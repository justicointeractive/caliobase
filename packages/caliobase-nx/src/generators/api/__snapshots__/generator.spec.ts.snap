// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api generator should run successfully: apps/api/project.json 1`] = `
"{
  \\"name\\": \\"api\\",
  \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
  \\"sourceRoot\\": \\"apps/api/src\\",
  \\"projectType\\": \\"application\\",
  \\"targets\\": {
    \\"build\\": {
      \\"executor\\": \\"@nrwl/webpack:webpack\\",
      \\"outputs\\": [
        \\"{options.outputPath}\\"
      ],
      \\"options\\": {
        \\"target\\": \\"node\\",
        \\"compiler\\": \\"tsc\\",
        \\"outputPath\\": \\"dist/apps/api\\",
        \\"main\\": \\"apps/api/src/main.ts\\",
        \\"tsConfig\\": \\"apps/api/tsconfig.app.json\\",
        \\"assets\\": [
          \\"apps/api/src/assets\\"
        ],
        \\"tsPlugins\\": [
          \\"@nestjs/swagger/plugin\\"
        ]
      },
      \\"configurations\\": {
        \\"production\\": {
          \\"optimization\\": true,
          \\"extractLicenses\\": true,
          \\"inspect\\": false,
          \\"fileReplacements\\": [
            {
              \\"replace\\": \\"apps/api/src/environments/environment.ts\\",
              \\"with\\": \\"apps/api/src/environments/environment.prod.ts\\"
            }
          ]
        }
      }
    },
    \\"serve\\": {
      \\"executor\\": \\"@nrwl/js:node\\",
      \\"options\\": {
        \\"buildTarget\\": \\"api:build\\"
      },
      \\"configurations\\": {
        \\"production\\": {
          \\"buildTarget\\": \\"api:build:production\\"
        }
      }
    },
    \\"lint\\": {
      \\"executor\\": \\"@nrwl/linter:eslint\\",
      \\"outputs\\": [
        \\"{options.outputFile}\\"
      ],
      \\"options\\": {
        \\"lintFilePatterns\\": [
          \\"apps/api/**/*.ts\\"
        ]
      }
    },
    \\"test\\": {
      \\"executor\\": \\"@nrwl/jest:jest\\",
      \\"outputs\\": [
        \\"coverage/apps/api\\"
      ],
      \\"options\\": {
        \\"jestConfig\\": \\"apps/api/jest.config.ts\\",
        \\"passWithNoTests\\": true
      }
    }
  },
  \\"tags\\": []
}
"
`;

exports[`api generator should run successfully: apps/api/src/app/app.module.ts 1`] = `
"import { SendRawEmailCommand, SES } from '@aws-sdk/client-ses';
import {
  CaliobaseModule,
  parseEnvVars,
  S3ObjectStorageProvider,
  TOKEN_TOKEN,
} from '@caliobase/caliobase';
import { DynamicModule, Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { createTransport } from 'nodemailer';
import * as SESTransport from 'nodemailer/lib/ses-transport';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { OrganizationProfile } from './entities/organization-profile.entity';
import { UserProfile } from './entities/user-profile.entity';


const ses = new SES({});

const envVars = parseEnvVars(({ required, optional }) => ({
  pgConnectionString:
    optional(
      'PG_CONNECTION_JSON',
      (v) => {
        const { host, port, username, password } = JSON.parse(v);
        return \`postgresql://\${username}:\${password}@\${host}:\${port}/postgres\`;
      },
      null
    ) ?? required('PG_CONNECTION_STRING', String),
  cmsHostname: required('CMS_HOSTNAME', String),
  emailFromAddress: required('EMAIL_FROM_ADDRESS', String),
  s3: {
    bucket: optional('S3_BUCKET', String, ''),
    keyPrefix: optional('S3_KEY_PREFIX', String, ''),
    endpoint: optional('S3_ENDPOINT', String, undefined),
  },
  cdn: {
    urlPrefix: optional('STATIC_FILE_BASEURL', String, ''),
  },
}));

@Module({
  imports: [],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {
  static async forRootAsync(): Promise<DynamicModule> {
    return {
      module: AppModule,
      providers: [],
      imports: [
        TypeOrmModule.forRoot({
          type: 'postgres',
          url: envVars.pgConnectionString,
          autoLoadEntities: true,
        }),
        
        CaliobaseModule.forRootAsync({
          urls: {
            forgotPassword: \`https://\${envVars.cmsHostname}/forgotPassword?token=\${TOKEN_TOKEN}\`,
          },
          controllerEntities: [],
          otherEntities: [],
          profileEntities: {
            OrganizationProfile,
            UserProfile,
            socialProfileToUserProfile: null,
          },
          objectStorageProvider: new S3ObjectStorageProvider({
            ...envVars.s3,
            cdnUrlPrefix: envVars.cdn.urlPrefix,
          }),
          emailTransport: createTransport(
            <SESTransport.Options>{
              SES: { aws: { SendRawEmailCommand }, ses },
            },
            { from: envVars.emailFromAddress }
          ),
        }),
      ],
    };
  }
}
"
`;

exports[`api generator should run successfully: apps/api/src/app/entities/organization-profile.entity.ts 1`] = `
"import {
  AbstractOrganizationProfile,
  CaliobaseEntity,
} from '@caliobase/caliobase';
import { Column } from 'typeorm';

@CaliobaseEntity({
  controller: { name: 'organization_profile' },
})
export class OrganizationProfile extends AbstractOrganizationProfile {
  @Column()
  name!: string;
}
"
`;

exports[`api generator should run successfully: apps/api/src/app/entities/user-profile.entity.ts 1`] = `
"import { AbstractUserProfile, CaliobaseEntity } from '@caliobase/caliobase';
import { Column } from 'typeorm';

@CaliobaseEntity({
  controller: { name: 'user_profile' },
})
export class UserProfile extends AbstractUserProfile {
  @Column()
  firstName!: string;

  @Column()
  lastName!: string;
}
"
`;

exports[`api generator should run successfully: apps/api/src/main.ts 1`] = `
"import { CaliobaseModule } from '@caliobase/caliobase';
import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';

async function bootstrap() {
  const app = await NestFactory.create(AppModule.forRootAsync(), {
    cors: true,
  });

  const globalPrefix = 'api';
  app.setGlobalPrefix(globalPrefix);

  await CaliobaseModule.bootstrap(app, {
    migration: {
      migrationsDir: './apps/api/migrations',
    },
  });

  const port = process.env.PORT || 3324;
  await app.listen(port);
  Logger.log(
    \`ðŸš€ Application is running on: http://localhost:\${port}/\${globalPrefix}\`
  );
}

bootstrap().catch((err) => {
  console.error(err);
  process.exit(1);
});
"
`;

exports[`api generator should run successfully: package.json 1`] = `
"{
  \\"name\\": \\"test-name\\",
  \\"dependencies\\": {
    \\"@nestjs/common\\": \\"^9.0.0\\",
    \\"@nestjs/core\\": \\"^9.0.0\\",
    \\"@nestjs/platform-express\\": \\"^9.0.0\\",
    \\"nodemailer\\": \\"^6.7.5\\",
    \\"reflect-metadata\\": \\"^0.1.13\\",
    \\"rxjs\\": \\"^7.0.0\\",
    \\"tslib\\": \\"^2.3.0\\"
  },
  \\"devDependencies\\": {
    \\"@nestjs/schematics\\": \\"^9.0.0\\",
    \\"@nestjs/testing\\": \\"^9.0.0\\",
    \\"@nrwl/eslint-plugin-nx\\": \\"14.7.18\\",
    \\"@nrwl/jest\\": \\"14.7.18\\",
    \\"@nrwl/linter\\": \\"14.7.18\\",
    \\"@nrwl/nest\\": \\"14.7.18\\",
    \\"@nrwl/node\\": \\"14.7.18\\",
    \\"@types/jest\\": \\"28.1.1\\",
    \\"@types/node\\": \\"16.11.7\\",
    \\"@types/nodemailer\\": \\"^6.4.4\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^5.36.1\\",
    \\"@typescript-eslint/parser\\": \\"^5.36.1\\",
    \\"eslint\\": \\"~8.15.0\\",
    \\"eslint-config-prettier\\": \\"8.1.0\\",
    \\"jest\\": \\"28.1.1\\",
    \\"jest-environment-jsdom\\": \\"28.1.1\\",
    \\"ts-jest\\": \\"28.0.5\\",
    \\"ts-node\\": \\"10.9.1\\"
  }
}
"
`;
